extend Cholesky decomposition method for finite fields

jacksonwalters opened the pull request:
Given a Hermitian matrix $U$ over $GF(q^2)$, returns a matrix $A$ such that $U = AA^\ast$, where $^\ast$ denotes conjugate-transpose.

Note that the Cholesky decomposition is meant for characteristic zero, and also only returns a lower/upper triangular matrix.

- Corresponding: [Issue #39202](https://github.com/sagemath/sage/issues/39202)
- Discussion: [Issue #38456](https://github.com/sagemath/sage/issues/38456)
- Required by: [PR #38455](https://github.com/sagemath/sage/pull/38455)

<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

jacksonwalters committed:
Update src/sage/matrix/matrix2.pyx

add hermitian_decomposition method

given a Hermitian matrix U, returns a matrix A such that U = AA*

use translated GAP source code for BaseChangeCanonical

raise ValueError for non full rank matrices

this method does not work for singular matrices. since we are already computing the rank in the row reduction, we can just check at the end whether the rank is full, and if not exit with a ValueError. this may be overcautious - it's not clear that this won't work for some singular matrices. there are cases where it certainly doesn't work, and we include one in the doctest.

add example for singular case

this example is currently failing and needs to be caught

remove unnecessary import, singularity check

if the matrix `U` is singular, the process will still result in a matrix but it will fail to have the expected property, namely `B*B.H == U`.

Update src/sage/matrix/matrix2.pyx

avoid import, directly call `sqrt` method, don't extend to keep result in `ZZ`

fix example

this was using the output from the `forms` package, when it should be from the GAP source translation.

change examples

there is a failing example due to the fact that the matrix is singular with q=3, and U = matrix(F,[[1,4,7],[4,1,4],[7,4,1]]).

move _cholesky_extended_ff to hidden method

move _cholesky_extended_ff to a separate hidden method, and just call it from inside cholesky if extended=true. add doctests and docs for the method separately. sparse for now, should be expanded upon.

Update src/sage/matrix/matrix2.pyx

add output for GF(3**2) case

this case fails to have a lower triangular decomposition which has been checked by brute force.

Update src/sage/matrix/matrix2.pyx

move two examples over finite fields from tests

the two examples computing the extended Cholesky decomposition over square order finite fields are in TESTS, and should be moved to EXAMPLES.

Update src/sage/matrix/matrix2.pyx

this is better and clarifies that the result of the Cholesky decomposition is lower triangular (and not possibly upper triangular).

Update src/sage/matrix/matrix2.pyx

add single whitespace before `extended` parameter

Update src/sage/matrix/matrix2.pyx

looks good. makes sense to point out the Cholesky decomposition might not exist, given that was our initial difficulty.

3 trailing whitespace

move hermitian_decomposition into cholesky

this method for decomposition a Hermitian matrix U = AA* is similar in spirit to the Cholesky decomposition, but extends it to work over finite fields of square order.

add two examples

add two examples of a Hermitian decomposition, one of which is not upper/lower triangular (so would be impossible with Cholesky decomposition)

Update src/sage/matrix/matrix2.pyx

yes, just forgot to update this

initialize row to -1

Update src/sage/matrix/matrix2.pyx

add word "package" to GAP ``forms``

Update src/sage/matrix/matrix2.pyx

change \ast to * for consistency

Trigger CI tests

use self.__class__ within matrix.pyx file

Trigger CI tests

remove blank line

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

Update src/sage/matrix/matrix2.pyx

use Matrix

handle 1x1 case

format failing test case properly

format failing test case properly by removing variables that hold results of computation. just call the function that raises the exception

Co-Authored-By: Travis Scrimshaw <clfrngrown@aol.com>

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+
+        ALGORITHM:
+
+        The algorithm is a translation of BaseChangeToCanonical in the GAP `forms` package for a Hermitian form.
```
```suggestion
        A conjugate-symmetric version of Gaussian elimination.
        This is a translation of ``BaseChangeToCanonical`` from
        the GAP ``forms`` for a Hermitian form.
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            sage: A == B*B.H
+            True
+
+        """
```
```suggestion
            sage: A = matrix(GF(11**2),[[1,4,7],[4,1,4],[7,4,1]])
            sage: B = A.hermitian_decomposition()
            sage: A == B * B.H
            True

        Verify we can decompose a Hermitian matrix for which
        there is no triangular decomposition::

            sage: A = matrix(GF(3**2), [[0, 1, 2], [1, 0, 1], [2, 1, 0]])
            sage: B = A.hermitian_decomposition()
            sage: A == B * B.H
            True
        """
```
Some formatting things.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+        n = self.nrows()
+
+        if not (F.is_finite() and F.order().is_square()):
+            raise ValueError("The base ring must be a finite field GF(q**2).")
```
```suggestion
            raise ValueError("the base ring must be a finite field of square order")
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+        r"""
+        Return the conjugate square root factorization of a Hermitian matrix.
+
+        Generally the base ring needs to be a finite field `GF(q**2)`.
```
```suggestion
        .. NOTE::

            The base ring needs to be a finite field `\GF{q^2}`.
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            k = u.log(z)
+            if k % (q+1) != 0:
+                raise ValueError("Unable to factor: u is not in base field GF(q)")
+            return z**(k // (q+1))
```
```suggestion
            return z ** (k//(q+1))
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            z = F.multiplicative_generator()
+            k = u.log(z)
+            if k % (q+1) != 0:
+                raise ValueError("Unable to factor: u is not in base field GF(q)")
```
```suggestion
                raise ValueError(f"unable to factor: {u} is not in base field GF({q})")
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            return z**(k // (q+1))
+
+        if not self.is_hermitian():
+            raise ValueError("matrix is not Hermitian")
```
```suggestion
```
Moving this earlier.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+        """
+        from sage.matrix.constructor import identity_matrix
+        from sage.misc.functional import sqrt
+
```
```suggestion

        if not self.is_hermitian():
            raise ValueError("matrix is not Hermitian")

```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+                D.add_multiple_of_row(row - 1, row, b)
+
+            # Eliminate below-diagonal entries in the current column
+            a = -A[row - 1, row - 1]**(-1)
```
```suggestion
            a = ~(-A[row-1, row-1])
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+                A.swap_columns(row, k)
+                D.swap_rows(row, k)
+
+                b = A[row, row - 1]**(-1)
```
```suggestion
                b = ~A[row, row-1]
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            row += 1
+
+            # Look for a non-zero element on the main diagonal, starting from `row`
+            i = row - 1  # Adjust for zero-based indexing in Sage
```
It is probably better to make `row` 0-based.

jacksonwalters replied:
I think we can just get rid of `i`. I'll look tomorrow.

tscrim replied:
I don't think this is possible as we mutate `i` compared to `row`.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+            for i in range(row, n):
+                b = A[i, row - 1] * a
+                if not b.is_zero():
+                    A.add_multiple_of_column(i,row - 1, b**q)
```
```suggestion
                    A.add_multiple_of_column(i, row-1, b**q)
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+
+        # Count how many variables have been used
+        if row == n - 1:
+            if not A[n - 1, n - 1].is_zero():
```
```suggestion
            if A[n-1, n-1]:  # nonzero entry
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
 
+    def hermitian_decomposition(self):
+        r"""
+        Return the conjugate square root factorization of a Hermitian matrix.
```
```suggestion
        Return a conjugate-transpose factorization of a Hermitian matrix.
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+
+        .. MATH::
+
+            A = B^\ast B
```
```suggestion
            A = B^* B,
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,149 @@ cdef class Matrix(Matrix1):
+
+        OUTPUT:
+
+        For a matrix `A` the routine returns a matrix `B` such that,
```
```suggestion
        For a Hermitian matrix `A` the routine returns a matrix `B` such that,
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+
+            A = B^* B,
+
+        where `B^\ast` is the conjugate-transpose.
```
```suggestion
        where `B^*` is the conjugate-transpose.
```
Let's be consistent. ;)

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+
+        A conjugate-symmetric version of Gaussian elimination.
+        This is a translation of ``BaseChangeToCanonical`` from
+        the GAP ``forms`` for a Hermitian form.
```
```suggestion
        the GAP ``forms`` package for a Hermitian form.
```

tscrim reviewed (changes requested):
Just a few more details, modulo making `row` 0-based.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+        This is a translation of ``BaseChangeToCanonical`` from
+        the GAP ``forms`` package for a Hermitian form.
+
+        TESTS:
```
Please add some `EXAMPLES::` as well.

jacksonwalters replied:
I have added two examples.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+
+        .. MATH::
+
+            A = B^* B,
```
```suggestion
            A = B B^*,
```
The tests are actually testing this.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+        row = -1
+
+        # Diagonalize A
+        while True:
```
```suggestion
        while row != n - 1:
```

tscrim replied:
Why didn't you do this change with the other below? I think this change clarifies the code a lot.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+                    A.add_multiple_of_row(i, row, b)
+                    D.add_multiple_of_row(i, row, b)
+
+            if row == n - 1:
+                break
```
```suggestion
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,6 +12952,156 @@ cdef class Matrix(Matrix1):
+
+            The base ring needs to be a finite field `\GF{q^2}`.
+
+        OUTPUT:
```
```suggestion
            The base ring needs to be a finite field `\GF{q^2}`.

        .. SEEALSO::

            :meth:`cholesky()`

        OUTPUT:
```

tscrim reviewed (commented):
As I look into the Cholesky decomposition and its doc, I am thinking we should actually call this method `cholesky_extended()` for "extended Cholesky decomposition". Or possibly accessing this via `cholesky(extended=True)` (with the default being `False`), in which case this method would be called something like `_cholesky_extended_ff`. I am actually leaning towards the latter right now since it is the most natural. Sorry for a somewhat big change so far along.

This paper is also somewhat relevant: [journal version](https://doi.org/10.1216/rmj.2024.54.423); [arXiv version](https://arxiv.org/abs/2202.04012v1.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12998,6 +12998,15 @@ cdef class Matrix(Matrix1):
+        NOTE:
+
+        For extended=True, the base ring needs to be a finite field `\GF{q^2}`.
+
```
```suggestion
        Over finite fields, the Cholesky decomposition might
        not exist, but when the field has square order (i.e.,
        `\GF{q^2}`), then we can perform the conjugate-symmetric
        version of Gaussian elimination, but the resulting
        decomposition matrix `L` might not be lower triangular.

        This is a translation of ``BaseChangeToCanonical`` fro
        the GAP ``forms`` package (for a Hermitian form).
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12952,7 +12952,7 @@ cdef class Matrix(Matrix1):
             return subspace
 
-    def cholesky(self):
+    def cholesky(self,extended=False):
```
```suggestion
    def cholesky(self, extended=False):
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -13168,6 +13177,18 @@ cdef class Matrix(Matrix1):
             ...
             ValueError: matrix is not positive definite
 
+        Use the extended method. May result in non upper/lower triangular matrix::
```
```suggestion
        Here we use the extended decomposition, where the result
        may not be a lower triangular matrix::
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -13214,7 +13235,131 @@ cdef class Matrix(Matrix1):
+            sage: A = matrix(GF(3**2), [[0, 1, 2], [1, 0, 1], [2, 1, 0]])
+            sage: B = A.cholesky(extended=True)
+            sage: A == B * B.H
+            True
```
Move all of these to the examples block.
```suggestion
```

jacksonwalters replied:
These two examples have been moved to the examples block in commit [dba8e2cb5de89be3e3200a2693ea1c8663cf2800](https://github.com/jacksonwalters/sage/commit/dba8e2cb5de89be3e3200a2693ea1c8663cf2800).

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -13168,6 +13177,18 @@ cdef class Matrix(Matrix1):
+            sage: U.cholesky(extended=True)
+            [        1         0         0]
+            [        4  9*z2 + 2         0]
+            [        7 10*z2 + 1  3*z2 + 3]
```
```suggestion
            sage: U = matrix(GF(5**2),[[0,1],[1,0]])
            sage: B = U.cholesky(extended=True); B
            [3*z2 4*z2]
            [3*z2   z2]
            sage: U == B * B.H
            True
            sage: U = matrix(GF(11**2),[[1,4,7],[4,1,4],[7,4,1]])
            sage: B = U.cholesky(extended=True); B
            [        1         0         0]
            [        4  9*z2 + 2         0]
            [        7 10*z2 + 1  3*z2 + 3]
            sage: U == B * B.H
            True
            sage: U = matrix(GF(3**2), [[0, 1, 2], [1, 0, 1], [2, 1, 0]])
            sage: B = A.cholesky(extended=True); B
            [[ADD OUTPUT]]
            sage: U == B * B.H
            True
```

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -13214,7 +13235,131 @@ cdef class Matrix(Matrix1):
+            sage: A == B * B.H
+            True
         """
+        def _cholesky_extended_ff():
```
Separate this out as a separate (hidden) method with its own documentation and doctest (different from the ones in the "main" method.

jacksonwalters replied:
This has been resolved in commit [b77a53eff3ef6b977caaecaa4b2782c65cc5b691](https://github.com/jacksonwalters/sage/commit/b77a53eff3ef6b977caaecaa4b2782c65cc5b691). 

When creating some new tests, I found a failing case (which also fails when using the `BaseChangeToCanonical` version from GAP `forms`). It is `q=3`, and `U = matrix(F,[[1,4,7],[4,1,4],[7,4,1]])`. This is because `U` is singular. 

I added a check to ensure the input is nonsingular. I don't know if this is overly restrictive.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -13214,7 +13235,131 @@ cdef class Matrix(Matrix1):
+            return D.inverse()
+
+        if extended:
+            return _cholesky_extended_ff()
```
```suggestion
        if extended:
            return self._cholesky_extended_ff()
```
So this will be the change.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12943,7 +12943,163 @@ cdef class Matrix(Matrix1):
+        if not (F.is_finite() and F.order().is_square()):
+            raise ValueError("the base ring must be a finite field of square order")
+
+        q = sqrt(F.order())
```
```suggestion
        q = F.order().sqrt(extend=False)
```
To avoid the import, directly call the applicable method, and guarantee the result remains in `ZZ`.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12943,7 +12943,163 @@ cdef class Matrix(Matrix1):
+            raise ValueError("matrix is not Hermitian")
+
+        if self.is_singular():
+            raise ValueError("matrix must be nonsingular")
```
```suggestion
```
I think this is somewhat expensive check that is unnecessary since it fails during the process. A division by 0 right? That would be the place to check (or perhaps catch the error).

Also, please add a doctest for the singular case (just in the helper method).

jacksonwalters replied:
That's the thing, it still produces an output matrix, it just doesn't satisfy the `B*B.H == U`. For example, `U = matrix(GF(3**2),[[1,4,7],[4,1,4],[7,4,1]])` yields `B = matrix(GF(3**2),[[1 0 0],[1 1 0],[1 0 1]])`. So we can only catch the error at the end, when it fails.

jacksonwalters replied:
Okay, I've added a (failing) singular example for now. I don't think there's a division by zero. It must be towards the end of the method where the rank is computed, and if it's not full rank it causes the issue.

jacksonwalters replied:
The above fail case is rank 1. The following fail case is rank 2: `U = matrix(F,[[0,4,7],[4,1,4],[7,4,1]])`. Note that at the end of the row reduction, we are already computing `r`, which is the rank of the input matrix. I propose we just check that the matrix is full rank, since it appears that it needs to be. Does it?

jacksonwalters replied:
No, it doesn't. The example `U = matrix(GF(3**2),[[0,0,0],[0,1,0],[0,0,1]])` should just return the same matrix. It returns `B = matrix(GF(3**2),[[0 0 1],[1 0 0],[0 1 0]])`, which does not satisfy `B*B.H == U`.

jacksonwalters replied:
It seems like in the initial `While` loop, it will note the first diagonal entry is zero, then try to do row/col swaps to ensure it's nonzero. I think for rank deficient matrices there may just be no way to do row/col swaps to get all diagonal entries nonzero. So we're left with a "bad" matrix, when really we shouldn't've done any swaps at all.

tscrim reviewed (changes requested):
Thank you. Just a few more little things. (It would be nice to squash some/all of the commits though since there are a lot of completely unhelpful commit messages...)

jacksonwalters committed:
fail cases should be tests, not examples

jacksonwalters committed:
add colon after Traceback

forgot the colon

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12943,7 +12943,177 @@ cdef class Matrix(Matrix1):
+                r = n - 1
+
+        if r < n:
+            raise ValueError("matrix is not full rank")
```
```suggestion
        if r < n:
            self.cache('rank', r)
            raise ValueError("matrix is not full rank")
```
Since you've gone through the trouble of computing the rank, let the matrix know that. :)

jacksonwalters replied:
Ah, slick, would not have thought to do that.

tscrim commented on src/sage/matrix/matrix2.pyx:
```diff
@@ -12943,7 +12943,177 @@ cdef class Matrix(Matrix1):
+
+        F = self._base_ring
+        n = self.nrows()
+
```
```suggestion
        n = self.nrows()
        if self.fetch("rank") not in [n, None]:
            raise ValueError("matrix is not full rank")
```
Here is a quick check if the rank is already known to the matrix.

tscrim reviewed (changes requested):
Thank you for everything. Two last little details to try and make it play better with the other methods.

jacksonwalters committed:
Update src/sage/matrix/matrix2.pyx

cache the rank of the matrix since are computing it anyways

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

jacksonwalters committed:
Update src/sage/matrix/matrix2.pyx

fetch the cached rank. if it's full or not yet computed, continue, otherwise throw a ValueError that the matrix is not full rank.

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

tscrim reviewed (approved):
I'm not sure about the mathematics off-hand. I'd say let's go with this for now, and we can come back to it later if there is a good use-case for it.

If the tests pass (morally), then I will set this to a positive review.
